proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,205 @@
+Fine-Grained Unused Warnings
+============================
+
+.. author:: Jakob Brünker
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/434>`_.
+.. contents::
+
+This proposal aims to make warnings about unused names more precise.
+
+Motivation
+----------
+
+::
+
+  a :: Int
+  a = 4
+    where bar = foo
+          foo = 3
+
+Compiling the above code with ``-Wall`` results in the
+these warnings:
+
+::
+
+      Defined but not used: `bar'
+      Defined but not used: `foo'
+
+Sometimes, I look at a warning like this and think "I could have sworn I used
+``foo`` somewhere" - but of course, what the warning is trying to say isn't
+actually that ``foo`` is unused, but rather that all of ``foo``'s usage sites
+are in unused bindings.
+
+Furthermore, while it's useful to see such a warning for CI, to immediately
+know all of the bindings that have to be removed to eliminate the warnings,
+during development, these additional warnings take up a lot of valuable
+screen real estate in ghcid.
+
+Proposed Change Specification
+-----------------------------
+The overarching idea is to separate warnings about names that are genuinely
+unused, and warnings that are only "transitively unused" (or "indirectly unused"), i.e. only used by
+other (transitively) unused bindings.
+
+The warnings about transitively unused names will be reported if a new flag, `-freport-indirectly-unsed-bindings` is enabled (by default, it is) *and* the appropriate existing warning flags about unused bindings is enabled, i.e. one of
```

Comment:
I actually did mean `indirectly-unused`. I've added that phrasing to alternatives, although I think I don't really like it, because I'm not sure what it means for a use to be indirect.